File: data/categories/fix_new_signature.py
import re

with open(r'c:\Users\Demch\OneDrive\projects\Cockatails-list\data\categories\new_signature.js', 'r', encoding='utf-8') as f:
    text = f.read()

# 1. glassware -> glass
text = re.sub(r'glassware:\s*', r'glass: ', text)

# 2. garnish objects -> strings
def repl_garnish(m):
    items = re.findall(r'name:\s*"([^"]+)"', m.group(1))
    if not items:
        return m.group(0)
    return 'garnish: "' + ' & '.join(items) + '"'

text = re.sub(r'garnish:\s*\[(.*?)\]', repl_garnish, text, flags=re.DOTALL)

# 3. ingredients quanties -> strings
def repl_ing(m):
    ing_text = m.group(1)
    
    def repl_qty(m2):
        name = m2.group(1)
        qty_str = m2.group(2)
        if qty_str == 'null':
            return f'name: "{name}", qty: "TBD"'
        try:
            q = float(qty_str)
            nl = name.lower()
            if 'bitter' in nl or 'tincture' in nl or 'foamer' in nl or 'solution' in nl or q <= 3:
                if 'solution' in nl or 'foamer' in nl:
                    suffix = ' pipette'
                else:
                    suffix = ' dashes' if q > 1 else ' dash'
                qs = str(int(q)) if q.is_integer() else str(q)
                return f'name: "{name}", qty: "{qs}{suffix}"'
            else:
                qs = str(int(q)) if q.is_integer() else str(q)
                return f'name: "{name}", qty: "{qs}ml"'
        except:
            return m2.group(0)
            
    ing_text = re.sub(r'name:\s*"([^"]+)",\s*qty:\s*([^}\s,]+)', repl_qty, ing_text)
    return 'ingredients: [' + ing_text + ']'

text = re.sub(r'ingredients:\s*\[(.*?)\]', repl_ing, text, flags=re.DOTALL)

# Remove totalQty
text = re.sub(r'\s*//.*?\n\s*totalQty:\s*[^,]+,\n', '\n', text, flags=re.DOTALL)
text = re.sub(r'\s*totalQty:\s*[^,]+,\n', '\n', text)

with open(r'c:\Users\Demch\OneDrive\projects\Cockatails-list\data\categories\new_signature.js', 'w', encoding='utf-8') as f:
    f.write(text)
print("Finished rewriting file")

File: data/categories/fix_new_signature.js
const fs = require('fs');

const filePath = 'c:\\Users\\Demch\\OneDrive\\projects\\Cockatails-list\\data\\categories\\new_signature.js';
let fileContent = fs.readFileSync(filePath, 'utf8');

// The file contains two identical arrays. Let's write a function to fix a list of cocktails
function fixCocktails(cocktails) {
  return cocktails.map(c => {
    // 1. Rename glassware to glass
    if (c.glassware) {
      c.glass = c.glassware;
      delete c.glassware;
    }

    // 2. Fix garnish
    if (Array.isArray(c.garnish)) {
      c.garnish = c.garnish.map(g => g.name).join(' & ');
    }

    // 3. Fix ingredients qty
    if (Array.isArray(c.ingredients)) {
      c.ingredients = c.ingredients.map(ing => {
        let qtyStr = ing.qty;
        if (typeof ing.qty === 'number') {
          const nameLower = ing.name.toLowerCase();
          if (nameLower.includes('bitter') || nameLower.includes('tincture') || nameLower.includes('foamer') || nameLower.includes('solution') || ing.qty <= 3) {
            qtyStr = ing.qty + (ing.qty > 1 ? " dashes" : " dash"); // default approximation
            if (nameLower.includes('solution') || nameLower.includes('foamer')) {
                qtyStr = ing.qty + " pipette";
            }
          } else {
            qtyStr = ing.qty + "ml";
          }
        } else if (ing.qty === null) {
          qtyStr = "TBD";
        }
        return { name: ing.name, qty: qtyStr };
      });
    }

    // remove totalQty if present
    if (c.totalQty !== undefined) {
      delete c.totalQty;
    }
    
    // Sort keys to look similar: id, name, category, method, glass, ice, garnish, ingredients
    const sorted = {};
    const order = ['id', 'name', 'category', 'method', 'glass', 'ice', 'garnish', 'ingredients'];
    order.forEach(k => {
      if (c[k] !== undefined) sorted[k] = c[k];
    });
    // add any remaining keys
    Object.keys(c).forEach(k => {
      if (!order.includes(k)) sorted[k] = c[k];
    });

    return sorted;
  });
}

// Extract the two arrays via regex or just eval the whole structure.
// Since it's a bit tricky to parse the JS programmatically and reconstruct it, 
// let's extract the array strings, evaluate them, format them, and stringify them.

const firstArrayMatch = fileContent.match(/window\.registerCocktails\('newSignature',\s*(\[\s*\{[\s\S]*?\}\s*\])\s*\);/);
const secondArrayMatch = fileContent.match(/data:\s*(\[\s*\{[\s\S]*?\}\s*\])\s*\}\);/);

if (firstArrayMatch && secondArrayMatch) {
  let firstArrayStr = firstArrayMatch[1];
  let secondArrayStr = secondArrayMatch[1];

  let firstArray = eval(firstArrayStr);
  let secondArray = eval(secondArrayStr);

  let fixedFirst = fixCocktails(firstArray);
  let fixedSecond = fixCocktails(secondArray);

  let newFirstStr = JSON.stringify(fixedFirst, null, 2);
  let newSecondStr = JSON.stringify(fixedSecond, null, 4);

  fileContent = fileContent.replace(firstArrayStr, newFirstStr);
  fileContent = fileContent.replace(secondArrayStr, newSecondStr);

  fs.writeFileSync(filePath, fileContent, 'utf8');
  console.log("Successfully fixed the cocktail structures.");
} else {
  console.log("Failed to match the arrays in the file.");
}
